cache/src/test/kotlin/com/dropbox/android/external/cache4/KeyedSynchronizerTest.kt

```diff
@@ -48,32 +40,27 @@ class KeyedSynchronizerTest {
             }
         }
 
-        delay(10)
+        delay(20)
 
-        // action 1 starts immediately
-        assertThat(action1Started).isTrue()
-        assertThat(action2Started).isFalse()
-        assertThat(action3Started).isFalse()
+        // One action has started
+        assertEquals(1, countTrue(action1Started, action2Started, action3Started))
```

Comment:
Can we assert a specific one has started, rather than any one of the 3?

Reply:
If you start three thread at the same time, then you can't tell which one starts executing first. So any of the three can start first.

Reply:
Right. If we add some `delays` between those launches we can expect them to be executed in order right (like the `a new action is queued after existing blocked actions using the same key from different thread` below)? I'm not too concerned with this one as it's not testing the order. But some the other tests focus more on execution order.

Reply:
The order of execution is not important for this test. The important thing is that the actions are executed sequentially. And the test shows that, even with random order of execution. So I think this is actually better than constructing a certain order of execution.

Reply:
Yes order it's not important for this test, but it is for the other 2 tests:

- `a new action is queued after existing blocked actions using the same key from different thread`
- `the next blocked action is unblocked when an action using the same key from another thread throws an exception`

as the 3rd action is only launched **after** 1st action is complete so execution order should be guaranteed. By doing those `delay(actionTime + 10` between assertions we are specifically letting the next expected action to start. But by using `countTrue`  for all the assertions, the tests will still pass even if action 2 starts before action 3, which is unexpected.

Reply:
In those two tests, if action1 gets the lock first, then the order should be action1, action2, action3. But if action2 gets the lock first, then the order should be action2, action1, action3.
If you want the order to be 1,2,3 then we have to make sure action1 gets the lock first and the thread for action2 starts before action1 finishes.

Reply:
Will do!